core/src/main/java/io/grpc/internal/AbstractServerStream.java

```diff
@@ -149,8 +154,11 @@ protected void inboundHeadersReceived(Metadata headers) {
       }
     }
     if (headers.containsKey(GrpcUtil.MESSAGE_ACCEPT_ENCODING_KEY)) {
-      pickCompressor(
+      Compressor c = pickCompressor(
           ACCEPT_ENCODING_SPLITER.split(headers.get(GrpcUtil.MESSAGE_ACCEPT_ENCODING_KEY)));
+      if (c != null) {
```

Comment:
> Since this is in the stream, it /should/ only get one set of headers, and it only live as long as the connection.

This needs to documented and enforced. Is `Stream.pickCompressor()` supposed to be called only once too?

> Now that you bring this up, can writeHeaders() be called multiple times, such as for trailers? If so, it wouldn't make sense to store it again or write it to the output headers.

Per javadoc, `ServerStream.writeHeaders()` can only be called once. You are right, `messageEncoding` is not supposed to be changed once it's set by the first set of headers.